File: ros_ws/src/projects/active_vision/scripts/toolViewPointCalc.py
#!/usr/bin/env python2

import numpy as np

minAngle = 20*(np.pi/180.0);

def sphericalToCartesian(pose,centre=[0,0,0]):
    res = [0,0,0]
    res[0] = centre[0]+pose[0]*np.sin(pose[2])*np.cos(pose[1])
    res[1] = centre[1]+pose[0]*np.sin(pose[2])*np.sin(pose[1])
    res[2] = centre[2]+pose[0]*np.cos(pose[2])
    return np.asarray(res)

def cartesianToSpherical(pose, centre=[0,0,0]):
  res = [0,0,0]
  res[0] = np.sqrt(pow(pose[0]-centre[0],2)+pow(pose[1]-centre[1],2)+pow(pose[2]-centre[2],2))
  res[1] = np.arctan2(pose[1]-centre[1],pose[0]-centre[0])
  res[2] = np.arctan2(np.sqrt(pow(pose[0]-centre[0],2)+pow(pose[1]-centre[1],2)),pose[2]-centre[2])
  return np.asarray(res)

def calcExplorationPoseB(startPose, dir):
  xyPlane = np.array([0,0,1]);

  stPoint = sphericalToCartesian(startPose);

  zAxis = stPoint; zAxis = zAxis/np.linalg.norm(zAxis);
  xAxis = np.cross(zAxis,xyPlane); xAxis /= np.linalg.norm(xAxis);
  yAxis = np.cross(zAxis,xAxis)

  switcher = {
        1: [ 1, 0],
        2: [ 1,-1],
        3: [ 0,-1],
        4: [-1,-1],
        5: [-1, 0],
        6: [-1, 1],
        7: [ 0, 1],
        8: [ 1, 1],
    }

  ratio = switcher.get(dir,[0,0])

  rotAxis = ratio[0]*xAxis + ratio[1]*yAxis; rotAxis /= np.linalg.norm(rotAxis)
  rotAxis.shape = (3,1)
  matA = np.array([[1,0,0],
                   [0,1,0],
                   [0,0,1]])
  matB = np.array([[0,-rotAxis[2],rotAxis[1]],
                   [rotAxis[2],0,-rotAxis[0]],
                   [-rotAxis[1],rotAxis[0],0]])
  matC = np.matmul(rotAxis,np.transpose(rotAxis))

  tempMat = np.cos(minAngle)*matA + np.sin(minAngle)*matB + (1-np.cos(minAngle))*matC
  endPoint = np.matmul(tempMat,stPoint)

  end = cartesianToSpherical(endPoint);

  # Polar angle 0 to 90 degree
  if(end[2] < 0):
    end[2] = -1*end[2]
    end[1] = end[1] + np.pi

  # Azhimuthal angle 0 to 360 degree
  end[1] = np.fmod(end[1],2*np.pi)
  if(end[1] < 0):
      end[1] += 2*np.pi

  return end.ravel()

def findDirection(start, end):
    diff = []
    for i in range(1,9):
        temp = np.asarray(calcExplorationPoseB(start,i))
        diff.append(round(np.sum(abs(end-temp)),2))
    min = np.amin(diff)
    idx = np.argmin(diff)
    if(min != 0):
        return -1
    else:
        return idx+1

File: ros_ws/src/projects/active_vision/src/toolViewPointCalc.cpp
#include <active_vision/toolViewPointCalc.h>

pcl::PointXYZ sphericalToCartesian(std::vector<double> &pose, pcl::PointXYZ &centre = defautCentre){
  pcl::PointXYZ res;
  res.x = centre.x+pose[0]*sin(pose[2])*cos(pose[1]);
  res.y = centre.y+pose[0]*sin(pose[2])*sin(pose[1]);
  res.z = centre.z+pose[0]*cos(pose[2]);
  return(res);
}

std::vector<double> cartesianToSpherical(pcl::PointXYZ &point, pcl::PointXYZ &centre = defautCentre){
  std::vector<double> res = {0,0,0};
  res[0] = sqrt(pow(point.x-centre.x,2)+pow(point.y-centre.y,2)+pow(point.z-centre.z,2));
  res[1] = atan2(point.y-centre.y,point.x-centre.x);
  res[2] = atan2(sqrt(pow(point.x-centre.x,2)+pow(point.y-centre.y,2)),point.z-centre.z);
  return(res);
}

bool checkValidPose(std::vector<double> pose){
	return (pose[2] <= 86*M_PI/180);
}

bool checkIfNewPose(std::vector<std::vector<double>> &oldPoses, std::vector<double> &pose, int type){
  for(int i = 0; i < oldPoses.size(); i++){
		if(type == 1){
			if(oldPoses[i] == pose) return false;
		}else{
			if(disBtwSpherical(oldPoses[i],pose) <= 0.25*(pose[0])*(20*M_PI/180)) return false;
		}
  }
  return true;
}

double disBtwSpherical(std::vector<double> &poseA,std::vector<double> &poseB){
	pcl::PointXYZ poseAcart = sphericalToCartesian(poseA);
	pcl::PointXYZ poseBcart = sphericalToCartesian(poseB);
	return(sqrt(pow(poseAcart.x-poseBcart.x,2)+pow(poseAcart.y-poseBcart.y,2)+pow(poseAcart.z-poseBcart.z,2)));
}

std::vector<double> calcExplorationPoseA(std::vector<double> &startPose, int dir, double step){
	double azimuthalOffset, polarOffset;
	switch(dir){
		case 1: azimuthalOffset = 0; 		 polarOffset = -step; break;	//N
		case 2: azimuthalOffset = step;  polarOffset = -step; break;	//NE
		case 3: azimuthalOffset = step;  polarOffset = 0; 		break;	//E
		case 4: azimuthalOffset = step;  polarOffset = step;  break;	//SE
		case 5: azimuthalOffset = 0; 		 polarOffset = step;  break;	//S
		case 6: azimuthalOffset = -step; polarOffset = step;  break;	//SW
		case 7: azimuthalOffset = -step; polarOffset = 0; 		break;	//W
		case 8: azimuthalOffset = -step; polarOffset = -step; break;	//NW
		default: printf("ERROR in calcExplorationPose\n");
	}

	std::vector<double> potentialPose = {startPose[0], startPose[1]+azimuthalOffset, startPose[2]+polarOffset};

	// Addressing the NW & NE scenario when polar angle goes from *ve to -ve
	if(potentialPose[2] < 0 && startPose[2] > 0) potentialPose[1] = startPose[1]-azimuthalOffset;

	// Polar angle should be +ve
	if(potentialPose[2] < 0){
		potentialPose[2] = -1*potentialPose[2];
		potentialPose[1] = potentialPose[1] + M_PI;
	}

	// Azhimuthal angle 0 to 360 degree
	potentialPose[1] = fmod(potentialPose[1],2*M_PI);
	if(potentialPose[1] < 0) potentialPose[1] += 2*M_PI;

	// std::cout << dir << " " << startPose[1]*180/M_PI << "," << startPose[2]*180/M_PI << "->" <<
	//                            potentialPose[1]*180/M_PI << "," << potentialPose[2]*180/M_PI << std::endl;

	return(potentialPose);
}

std::vector<double> calcExplorationPoseB(std::vector<double> &startPose, int dir, double step){

  Eigen::Vector3f xAxis,yAxis,zAxis,rotAxis,tempVec;
  Eigen::Vector3f xyPlane(0,0,1);
  Eigen::Matrix3f matA; matA << 1,0,0,0,1,0,0,0,1;
  Eigen::Matrix3f matB, matC, tempMat;
  // tf::Matrix3x3 rotMat;

  pcl::PointXYZ centre(0,0,0);
  pcl::PointXYZ stPoint = sphericalToCartesian(startPose);
  pcl::PointXYZ endPoint;

  zAxis = stPoint.getVector3fMap(); zAxis.normalize();
  xAxis = zAxis.cross(xyPlane); xAxis.normalize();
  yAxis = zAxis.cross(xAxis);

	std::vector<double> ratio={0,0};
	switch(dir){
		case 1: ratio[0]=+1; ratio[1]=0;   break;
		case 2: ratio[0]=+1; ratio[1]=-1;  break;
		case 3: ratio[0]=0;  ratio[1]=-1;  break;
		case 4: ratio[0]=-1; ratio[1]=-1;  break;
		case 5: ratio[0]=-1; ratio[1]=0;   break;
		case 6: ratio[0]=-1; ratio[1]=+1;  break;
		case 7: ratio[0]=0;  ratio[1]=+1;  break;
		case 8: ratio[0]=+1; ratio[1]=+1;  break;
	}

  rotAxis = ratio[0]*xAxis + ratio[1]*yAxis; rotAxis.normalize();
  matB << 0,-rotAxis[2],rotAxis[1],rotAxis[2],0,-rotAxis[0],-rotAxis[1],rotAxis[0],0;
  matC = rotAxis*rotAxis.transpose();

  tempMat = cos(step)*matA + sin(step)*matB + (1-cos(step))*matC;
  tempVec = tempMat*stPoint.getVector3fMap();
  endPoint.x = tempVec[0]; endPoint.y = tempVec[1]; endPoint.z = tempVec[2];

  std::vector<double> end = cartesianToSpherical(endPoint);

  // Polar angle 0 to 90 degree
  if(end[2] < 0){
    end[2] = -1*end[2];
    end[1] = end[1] + M_PI;
  }

  // Azhimuthal angle 0 to 360 degree
  end[1] = fmod(end[1],2*M_PI);
  if(end[1] < 0) end[1] += 2*M_PI;

	return end;
}

std::vector<double> calcExplorationPose(std::vector<double> &startPose, int dir, int mode, double step){
  switch(mode){
    case 1: return calcExplorationPoseA(startPose,dir,step); break;
    case 2: return calcExplorationPoseB(startPose,dir,step); break;
    default: return calcExplorationPoseA(startPose,dir);
  }
}
